### ISSUE TITLE
[BUG] CI unable to resolve flutter command after successful fvm install

### ISSUE THREAD

**Issue by ened** [leoafarias/fvm#160]
**Before creating a bug report please make check the following**

**Describe the bug**

fvm works in interactive shell but fails to find `flutter` command internally when running on CI

**To Reproduce**
1. Start a CI job on Gitlab and include the following

```
install_flutter:
  stage: pre-build
  script:
    - export PATH="$PATH":"$HOME/.pub-cache/bin"
    - fvm --verbose install 1.20.4
  tags:
    - macos
```

2. This fails on the CI with this error:

```
22Version: 1.20.4 - already installed.
23Internal Error: Could not finish setting up Flutter sdk
24#0      flutterSetupWorkflow (package:fvm/src/workflows/flutter_setup.workflow.dart:9:5)
25<asynchronous suspension>
26#1      InstallCommand.run (package:fvm/src/commands/install_command.dart:49:13)
27<asynchronous suspension>
28#2      CommandRunner.runCommand (package:args/command_runner.dart:197:27)
29#3      CommandRunner.run.<anonymous closure> (package:args/command_runner.dart:112:25)
30#4      new Future.sync (dart:async/future.dart:223:31)
31#5      CommandRunner.run (package:args/command_runner.dart:112:14)
32#6      fvmRunner (package:fvm/src/runner.dart:34:23)
33#7      main (file:///Users/build/.pub-cache/hosted/pub.dartlang.org/fvm-1.3.0/bin/main.dart:4:9)
34#8      _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:299:32)
35#9      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:168:12)
36Unhandled exception:
37Internal Error: Could not finish setting up Flutter sdk
38#0      flutterSetupWorkflow (package:fvm/src/workflows/flutter_setup.workflow.dart:9:5)
39<asynchronous suspension>
40#1      InstallCommand.run (package:fvm/src/commands/install_command.dart:49:13)
41<asynchronous suspension>
42#2      CommandRunner.runCommand (package:args/command_runner.dart:197:27)
43#3      CommandRunner.run.<anonymous closure> (package:args/command_runner.dart:112:25)
44#4      new Future.sync (dart:async/future.dart:223:31)
45#5      CommandRunner.run (package:args/command_runner.dart:112:14)
46#6      fvmRunner (package:fvm/src/runner.dart:34:23)
47#7      main (file:///Users/build/.pub-cache/hosted/pub.dartlang.org/fvm-1.3.0/bin/main.dart:4:9)
48#8      _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:299:32)
49#9      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:168:12)
```

**Expected behavior**
Should behave like in interactive session: 

```
% fvm install 1.20.4
Version: 1.20.4 - already installed.
Flutter 1.20.4 • channel unknown • unknown source
Framework • revision fba99f6cf9 (3 weeks ago) • 2020-09-14 15:32:52 -0700
Engine • revision d1bc06f032
Tools • Dart 2.9.2
```

**Desktop (please complete the following information):**
 - OS: [MacOS ]
 - FVM Version [1.3.0] via Homebrew
**Additional context**

**Comment by mx1up** [comment]
I can confirm this problem on windows and linux.
It has to do with the logging changes (stdin.echoMode and stdin.lineMode). Personally, I would simply always show the output, that way you see what's going on..

**Comment by leoafarias** [comment]
@mx1up are you using version 1.3.0 correct?

**Comment by mx1up** [comment]
yes, sorry, i mixed input/output up somehow :)

**Comment by leoafarias** [comment]
@mx1up I was not able to replicate this issue on the platforms specified. Could you give a bit more detail about the problem and when you see it?

**Comment by mx1up** [comment]
hi Leo, I'm running the install command on a version pinned in the json, the version is already installed (don't know if this is relevant). I will test more elaborately next week. the OP specifies the version on the command line, which seems to exhibit the same behavior.

**Comment by Clean-Cole** [comment]
I was able to reproduce this on Catalina in our CI.  I'm pretty sure I'm using the same CI shell executor system as @ened. 


- macOS Catalina 10.15.7 (19H2)
- GitLab Runner 13.4.1
- FVM v1.3.1

Only thing of note is that I have gone ahead and taken in the change to `zsh` from `bash` (gitlab runner still uses bash) but I am able to run all the `fvm` commands successfully in an interactive shell with `zsh` and `bash`.  Everything is correct in regards to my `$PATH` etc.

I'm taking a swing at trying to create a test to reproduce reliably.  Not sure how far I will make it down that path.

```
$ gitlab-runner --version
Version:      13.4.1
Git revision: e95f89a0
Git branch:   13-4-stable
GO version:   go1.15.2
Built:        2020-10-03T02:51:53+01:00
OS/Arch:      darwin/amd64
```

Here is another stack trace.  

```
$ pub global activate fvm
 Package fvm is currently active at version 1.3.1.
 Resolving dependencies...
 + args 1.6.0
 + charcode 1.1.3
 + cli_dialog 0.3.3
 + cli_util 0.2.0
 + collection 1.14.13
 + console 3.1.0
 + date_format 1.0.9
 + functional_data 0.3.0
 + fvm 1.3.1
 + http 0.12.2
 + http_parser 3.1.4
 + io 0.3.4
 + json2yaml 1.0.1
 + meta 1.2.3
 + path 1.7.0
 + pedantic 1.9.2
 + plain_optional 0.1.3
 + pretty_json 1.1.0
 + process_run 0.10.12+3 (0.11.0+2 available)
 + pub_semver 1.4.4
 + pubspec_yaml 2.0.1
 + source_span 1.7.0
 + string_scanner 1.0.5
 + sum_types 0.2.1+2
 + term_glyph 1.1.0
 + typed_data 1.2.0
 + vector_math 2.0.8
 + version 1.2.0
 + yaml 2.2.1
 Precompiling executables...
 Precompiled fvm:main.
 Installed executable fvm.
 Activated fvm 1.3.1.
 $ fvm --verbose install
 Version: 1.22.1 - already installed.
 Internal Error: Could not finish setting up Flutter sdk
 #0      flutterSetupWorkflow (package:fvm/src/workflows/flutter_setup.workflow.dart:9:5)
 <asynchronous suspension>
 #1      InstallCommand.run (package:fvm/src/commands/install_command.dart:51:13)
 <asynchronous suspension>
 #2      CommandRunner.runCommand (package:args/command_runner.dart:197:27)
 #3      CommandRunner.run.<anonymous closure> (package:args/command_runner.dart:112:25)
 #4      new Future.sync (dart:async/future.dart:223:31)
 #5      CommandRunner.run (package:args/command_runner.dart:112:14)
 #6      fvmRunner (package:fvm/src/runner.dart:34:23)
 #7      main (file:///Users/gitlabci/.pub-cache/hosted/pub.dartlang.org/fvm-1.3.1/bin/main.dart:4:9)
 #8      _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:299:32)
 #9      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:168:12)
 Unhandled exception:
 Internal Error: Could not finish setting up Flutter sdk
 #0      flutterSetupWorkflow (package:fvm/src/workflows/flutter_setup.workflow.dart:9:5)
 <asynchronous suspension>
 #1      InstallCommand.run (package:fvm/src/commands/install_command.dart:51:13)
 <asynchronous suspension>
 #2      CommandRunner.runCommand (package:args/command_runner.dart:197:27)
 #3      CommandRunner.run.<anonymous closure> (package:args/command_runner.dart:112:25)
 #4      new Future.sync (dart:async/future.dart:223:31)
 #5      CommandRunner.run (package:args/command_runner.dart:112:14)
 #6      fvmRunner (package:fvm/src/runner.dart:34:23)
 #7      main (file:///Users/gitlabci/.pub-cache/hosted/pub.dartlang.org/fvm-1.3.1/bin/main.dart:4:9)
 #8      _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:299:32)
 #9      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:168:12)
```

**Comment by leoafarias** [comment]
@Clean-Cole can you run 'fvm flutter --version' in the interactive shell and add the log here?

**Comment by Clean-Cole** [comment]
Here you go @leoafarias this is straight from the CI machine.

```
% fvm --verbose flutter --version
FVM: Running version 1.22.1
Flutter 1.22.1 • channel unknown • unknown source
Framework • revision f30b7f4db9 (4 days ago) • 2020-10-08 10:06:30 -0700
Engine • revision 75bef9f6c8
Tools • Dart 2.10.1
```

For what it's worth if I add the constructor back into the `ConsoleController` class everything seems to play nice again.  There isn't any other place in the code that seems to set that `isCli` in the master branch except for its initial value in the constants file.

```
ConsoleController() {
    isCli = io.stdin.hasTerminal;
}
```

**Comment by leoafarias** [comment]
@Clean-Cole This is changed in the runner.
https://github.com/leoafarias/fvm/blob/master/packages/cli/lib/src/runner.dart#L22

I will take a deeper look to understand this problem. Also will release a -dev version for better logging so we can see what is happening

**Comment by leoafarias** [comment]
@Clean-Cole I deployed version 1.3.2, and added some better tracing to see more info on that setup error

**Comment by Clean-Cole** [comment]
Nice, now we have a thread to pull on.  Cursory searches for the error bring up a lot of different CLI tools over the years having similar-ish issues.  One stuck out [here](https://github.com/pyinvoke/invoke/issues/425#issuecomment-297502679) 

tldr; from that guy is, maybe not calculating the whether the proc is atty or not before the a headless process begins might change the behavior?  I have full use of this build machine to continue testing if you want me to try anything else. 

```
 Installed executable fvm.
 Activated fvm 1.3.2.
 $ fvm --verbose install
 Version: 1.22.1 - already installed.
 [     0.421s] StdinException: Error setting terminal echo mode, OS Error: Inappropriate ioctl for device, errno = 25
 Internal Error: Could not finish setting up Flutter sdk
 #0      flutterSetupWorkflow (package:fvm/src/workflows/flutter_setup.workflow.dart:11:5)
 <asynchronous suspension>
 #1      InstallCommand.run (package:fvm/src/commands/install_command.dart:51:13)
 <asynchronous suspension>
 #2      CommandRunner.runCommand (package:args/command_runner.dart:197:27)
 #3      CommandRunner.run.<anonymous closure> (package:args/command_runner.dart:112:25)
 #4      new Future.sync (dart:async/future.dart:223:31)
 #5      CommandRunner.run (package:args/command_runner.dart:112:14)
 #6      fvmRunner (package:fvm/src/runner.dart:34:23)
 #7      main (file:///Users/gitlabci/.pub-cache/hosted/pub.dartlang.org/fvm-1.3.2/bin/main.dart:4:9)
 #8      _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:299:32)
 #9      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:168:12)
 Unhandled exception:
 Internal Error: Could not finish setting up Flutter sdk
 #0      flutterSetupWorkflow (package:fvm/src/workflows/flutter_setup.workflow.dart:11:5)
 <asynchronous suspension>
 #1      InstallCommand.run (package:fvm/src/commands/install_command.dart:51:13)
 <asynchronous suspension>
 #2      CommandRunner.runCommand (package:args/command_runner.dart:197:27)
 #3      CommandRunner.run.<anonymous closure> (package:args/command_runner.dart:112:25)
 #4      new Future.sync (dart:async/future.dart:223:31)
 #5      CommandRunner.run (package:args/command_runner.dart:112:14)
 #6      fvmRunner (package:fvm/src/runner.dart:34:23)
 #7      main (file:///Users/gitlabci/.pub-cache/hosted/pub.dartlang.org/fvm-1.3.2/bin/main.dart:4:9)
 #8      _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:299:32)
 #9      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:168:12)
```

Thanks for the quick and helpful responses, love this tool.

**Comment by mx1up** [comment]
Hi ,

I have not had time yet to investigate in detail, so apologies if my remark is irrelevant, but I found the following. I created the following small script:
```
import 'dart:io';

Future<void> main(List<String> arguments) async {
  print(stdioType(stdin));
}

```
and executed it in my CI environment (teamcity). This is the output:
`StdioType: pipe`

When I execute the script from a terminal, it says
`StdioType: terminal`
When I run the fvm install command from an ssh shell on the CI server, it works without trouble.

When running fvm dynamically from intellij, it also has the Pipe type, that's probably how i validated the problem on windows, so that's not really representative. Next, I tried latest fvm in command prompt (normal user and elevated), but both crash the command prompt (at the end of fvm execution), goes too fast to see if the other problem is gone. this crash is probably another problem though

**Comment by leoafarias** [comment]
Great info guys. This Dart API seems to be very sensitive... there is an issue that it causes errors even based on the order you call the properties, which seems less than ideal.

I created a dev version to test the fix, let me know if this works. Checking if there is a terminal attached also before calling it.

`pub global activate fvm 1.3.3-dev.0`

**Comment by Clean-Cole** [comment]
it works 🎉  but also the plot thickens a bit? 🤔 

So.. on the CI machine We make it past the `fvm install` command no problem. The very next command is running flutter tests. `fvm flutter test` which it fails.  Now the failure it self is actually not an fvm issue.  Someone broke our pubspec somewhere else. On our main pipeline box I have fvm pinned to v1.2.3 and I can see an error being returned in the console, but on this dev version the output is suppressed.

For me personally I can live without that, but I imagine this might be a blocker to other folks.

##### Output from typing in terminal on the machine (Shows expected output)
```
gitlabci@gitlabcirunner this-and-that % fvm install
Version: 1.22.1 - already installed.
Flutter 1.22.1 • channel unknown • unknown source
Framework • revision f30b7f4db9 (5 days ago) • 2020-10-08 10:06:30 -0700
Engine • revision 75bef9f6c8
Tools • Dart 2.10.1
gitlabci@gitlabcirunner this-and-that % fvm flutter test
FVM: Running version 1.22.1
Running "flutter pub get" in this-and-that...
Because every version of flutter_driver from sdk depends on process 4.0.0-nullsafety.2 which depends on platform ^3.0.0-nullsafety.1, every version of flutter_driver from sdk requires platform ^3.0.0-nullsafety.1.

And because every version of data from path depends on firebase_messaging ^6.0.9 which depends on platform ^2.0.0, flutter_driver from sdk is incompatible with data from path.

So, because flutter_app depends on both data from path and flutter_driver any from sdk, version solving failed.
pub get failed (1; So, because flutter_app depends on both data from path and flutter_driver any from sdk, version solving failed.)
```


###### Same Failure on CI Machine (no error output shown in  gitlab CI logs)
```
Downloading fvm 1.3.3-dev.0...
 Precompiling executables...
 Precompiled fvm:main.
 Installed executable fvm.
 Activated fvm 1.3.3-dev.0.
 $ fvm --verbose install
 Version: 1.22.1 - already installed.
 $ fvm --verbose flutter test
 FVM: Running version 1.22.1
 ERROR: Job failed: exit status 1
```

**Comment by leoafarias** [comment]
That's an easy fix I am using the terminal logic to determine stdout now, I will just go back to how it was before.

Will open an issue with the dart team in regards to the sdin API also.

**Comment by leoafarias** [comment]
@Clean-Cole  `pub global activate fvm 1.3.3-dev.1`

**Comment by mx1up** [comment]
I tested 1.3.3-dev.1 on linux CI and terminal. it works! :partying_face:

**Comment by ened** [comment]
I can confirm it as well. Thank you everyone for seeing this through. Really appreciate.
